multigeiger/speaker.cpp

```diff
@@ -68,8 +68,8 @@ void IRAM_ATTR isr_audio() {
   }
   portEXIT_CRITICAL_ISR(&mux_audio);
 
-  if (!isr_sequence)
-    return;  // nothing to do
+  if (!playing_audio && !playing_tick)
+    return;
```

Comment:
yeah, that should work.

Reply:
ehrm, guess that is equivalently bad as the isr_sequence reading we had before - both are variables changed by isr code.

so better revert to the "p" method.

Reply:
Then we have the issue with every solution, as also "p" is part of the isr_sequence commands. I propose now an else branch with a second EXIT mux_audio, as this seems the way to go. Any new variable will also be part of the isr_sequence, right?

Reply:
p is a local dynamic variable, so each user has its own copy of it.
the mux is only needed if multiple users access the same memory location (global or local static variable).

Reply:
Hmm, but in order to check for NULL, we have to lift its scope up to the isr_audio function. And according to [this description](https://www.arduino.cc/reference/de/language/variables/variable-scope-qualifiers/volatile/), volatile does not really help here. 
What about the if-else proposition?

Reply:
it will be still local and dynamically allocated within the isr_audio function, so it is neither used by other functions nor by other invocactions of isr_audio (not sure if that can happen).

that volatile description is likely meant for the 8bit arduino, so parts may not apply to esp32.

Reply:
btw, guess the volatile is not really needed here, but also does no harm.

Reply:
No, I still don't like it. 
We get `p == NULL` when the sequence reaches the end and the routine returns, but the last sequence part is not processed and the isr*_sequences are not reset, nor are the playing_ bools. So I opt 

- either for the current proposition with repeated `portEXIT `in if-else
- or for a new `portENTER - portEXIT` block around a 'playing_tick || playing_audio' if-loop

Both process the last sequence part correctly.
What would you prefer?

Reply:
...and the second option would also require to 'portEXIT's

Reply:
```
portENTER_CRITICAL_ISR(&mux_audio);
if (!playing_tick && !playing_audio) {
   portEXIT_CRITICAL_ISR(&mux_audio);
   return;
}
portEXIT_CRITICAL_ISR(&mux_audio);
```

Reply:
not sure what you mean / what the problem is.

If we assign to p within the mutually exclusive block, then no other mutually exclusive block (using the same mutex) will run at the same time. so especially the mutex block at the end will not have partially executed.

Reply:
I've commented in the original version, s. below